### PR TITLE
Stabilize if_let_rescope

### LINKED ISSUE #131154: Proposal: stabilize `if_let_rescope` for Edition 2024
Tracked by #124085

# Summary

`if_let_rescope` is an implementation of the #124085 through #107251. This change did not come with a RFC. This proposal will aim to describe the change to the language fully.

`if let` and the current unstabilized `if let`-chain has a caveat to the assignment of lifetime rules regarding temporary values generated from initializer expressions.

```rust=
//@ edition: 2021
if let Some(value) = droppy().get() {
    // do something
} else {
    // `droppy()` is still alive here
    // but it is out of reach as well.
}
```
Instead, after stabilizing this, we have this effect.
```rust=
//@ edition: 2024
#![feature(if_let_rescope)]
if let Some(value) = droppy().get() {
    // do something
} else {
    // `droppy()` is already dropped here
}
```

This will allow us to be more consistent with `let $pat = $expr else { $diverge; }` where temporaries from `$expr` are dropped before entering the `$diverge` branch in today's language.

Given that this is an Edition breaking change, a lint is developed and tested in the field with crater runs.

# What is being proposed for stabilization

In #107251, a new kind of scope data `IfThenRescope` is introduced and gated behind the edition and this feature gate. Upon stabilization, this scope will work like a terminating scope for variables and temporary values generated from the `let`-bindings.

What is more significant is the case where `if let` is used as a non-trivial value-producing subexpression in a larger evaluation context, in which it will have semantic impact to lifetimes and object drop orders.

Here are some breaking changes that will be introduced by stabilization.
```rust=
#![feature(if_let_rescope)]
// borrowck may reject this ...
run(if let Some(ref value) = droppy().borrow() {
    value
} else {
    something_else()
})
// ...
```
A marginal amount of crates are discovered by previous crater runs in which this change leads to rejection from the borrow checker, because `droppy()` does not live long enough.

The migration lint that is now implemented on `master` suggests a rewrite to restore the old semantics perfectly, but not automatically applicable due to the current limitation of machine applicable lints and our internal support for *overlapping suggestion spans*.

```rust=
#![feature(if_let_rescope)]
// borrowck will not reject this ...
run(match droppy().borrow() {
    Some(ref value) => { value }
    _ => { something_else() }
})
```

A more subtle and silent breakage could be involving use of types with significant drop implementation and synchronization primitives such as `Mutex`es.

```rust=
if let Some(value) = mutex.lock().unwrap().get() {
    // do something
} else {
    // somehow code expects the lock to be held here ...
    // or because the mutex was only improperly used as a semaphore or for signalling ...
}
```

Although the crater run did not find breakage in runtime behaviour through `cargo test`, we proceed with developing lints to detect those cases and rewrite them into `match`es. Here the suggestion notes are machine applicable because we can coalesce the overlapping spans in one pass.

In the latest [crater run](https://github.com/rust-lang/rust/pull/129466#issuecomment-2383141018) we found a marginal population of crates that are truly impacted by this lint, revealing corner cases with the lint which are now properly handled and tested.

# Future interactions

Currently there is another possible use of pattern matching in control flows that complements `if let` chain is the `is` operator proposed in rust-lang/rfcs#3573. This change both runs along the same principle of no unnecessary binding and long lifetime beyond the applicable scope from a pattern-matching predicate. With this change, the implementation of `is` would be believably less error-prone due to the unexpectedly long lifetimes.

This leaves `match` being the only exception to this principle. In fact, our mitigation and migration strategy is based on it. It might remains as an acceptable exception and further be publicized and inadvertently advertised as such.

In general, this change is not expected to influence or hinder major language design.

### PR DESCRIPTION
**Pull request by dingxiangfei2009**
Close #131154
Tracked by #124085

### CODE REVIEW

File: compiler/rustc_feature/src/accepted.rs [hunk 1]
```diff
@@ -244,6 +244,8 @@ declare_features! (
     (accepted, irrefutable_let_patterns, "1.33.0", Some(44495)),
     /// Allows `#[instruction_set(_)]` attribute.
     (accepted, isa_attribute, "1.67.0", Some(74727)),
+    /// Rescoping temporaries in `if let` to align with Rust 2024.
+    (accepted, if_let_rescope, "1.83.0", Some(124085)),
```
Comment (ehuss) [inline compiler/rustc_feature/src/accepted.rs:248]: Aside from being sorted, this should also probably use `CURRENT RUSTC VERSION`.

**Comment by traviscross** [comment]
r? @lcnr

Given that you had a look at the related...

- https://github.com/rust-lang/rust/pull/131983

...it may make sense for you to have a look at this also.

File: compiler/rustc_feature/src/accepted.rs (hunk 1 above)
Comment (dingxiangfei2009) [reply compiler/rustc_feature/src/accepted.rs]: Applied

**Review by lcnr — COMMENTED** [review]
r=me on the impl, not gonna actually approve this myself while I am arguing about the current design in #131154 :>

**Comment by Kobzol** [comment]
@rust-timer build 701efc4f27da7e80269be0f0d119a5e8aadfe54e

Checking if this might be the source of the regression in https://github.com/rust-lang/rust/pull/132326.

**Comment by Kobzol** [comment]
Yeah so this is quite the regression on instruction counts. Doesn't look so terrible on cycles though. The [detailed results](https://perf.rust-lang.org/detailed-query.html?commit=701efc4f27da7e80269be0f0d119a5e8aadfe54e&benchmark=unicode-normalization-0.1.19-check&scenario=incr-full&base_commit=e473783d90e2289b8a97575fa60d6315f0a318eb&sort_idx=-11) show the `lint_mod` query taking more time, could that be relevant? (it can also be something completely different though, or just an effect of this change in general).

**Comment by rylev** [comment]
Ping @dingxiangfei2009 @lcnr - the regression here is pretty large. Would it be worth prioritizing investigation (started by @Kobzol above) or even considering a partial revert?

**Comment by Kobzol** [comment]
I'm not sure how practical it is to revert this outright tbh, since it's heading into Edition 2024, AFAIK.

**Comment by traviscross** [comment]
Indeed not practical to revert, as you say, for that reason.  Would be better to understand and fix in place.

**Pull request merged**